libs/compact-map/test/Test/Compact/KeyMap.hs

```diff
@@ -25,9 +64,409 @@ instance Arbitrary a => Arbitrary (KeyMap a) where
+unionHMDATA n m = HM.union (HM.fromList n) (HM.fromList m) $==$ Map.union (Map.fromList n) (Map.fromList m)
+
+intersectHMDATA :: forall v. (Eq v, Show v) => [(Key, v)] -> [(Key, v)] -> Property
+intersectHMDATA n m = HM.intersection (HM.fromList n) (HM.fromList m) $==$ Map.intersection (Map.fromList n) (Map.fromList m)
+
+lookupHMDATA :: forall v. (Eq v, Show v) => Key -> [(Key, v)] -> Property
+lookupHMDATA k m = HM.lookup k (HM.fromList m) === Map.lookup k (Map.fromList m)
+
+restrictHMDATA :: (Show a, Eq a) => [(Key, a)] -> Set Key -> Property
+restrictHMDATA m s = HM.restrictKeys (HM.fromList m) s $==$ Map.restrictKeys (Map.fromList m) s
+
+withoutHMDATA :: (Show a, Eq a) => [(Key, a)] -> Set Key -> Property
+withoutHMDATA m s = HM.withoutKeys (HM.fromList m) s $==$ Map.withoutKeys (Map.fromList m) s
+
+minHMDATA :: (Eq a, Show a) => [(Key, a)] -> Property
+minHMDATA m = HM.lookupMin (HM.fromList m) === Map.lookupMin (Map.fromList m)
+
+maxHMDATA :: (Eq a, Show a) => [(Key, a)] -> Property
+maxHMDATA m = HM.lookupMax (HM.fromList m) === Map.lookupMax (Map.fromList m)
+
+splitHMDATA :: (PrettyA a, Eq a, Show a) => Key -> [(Key, a)] -> Property
+splitHMDATA k m = case Map.splitLookup k (Map.fromList m) of
+  (a, b, c) -> (fromList (Map.toList a), b, fromList (Map.toList c)) === splitLookup k (fromList m)
+
+minViewHMDATA :: [(Key, Int)] -> Property
+minViewHMDATA m = case Map.minViewWithKey (Map.fromList m) of
+  Nothing -> minViewWithKey (fromList m) === Nothing
+  Just (b, c) -> Just (b, fromList (Map.toList c)) === minViewWithKey (fromList m)
+
+maxViewHMDATA :: [(Key, Int)] -> Property
+maxViewHMDATA m = case Map.maxViewWithKey (Map.fromList m) of
+  Nothing -> maxViewWithKey (fromList m) === Nothing
+  Just (b, c) -> Just (b, fromList (Map.toList c)) === maxViewWithKey (fromList m)
+
+keyMapEquivDataMap :: TestTree
+keyMapEquivDataMap =
+  testGroup
+    "KeyMap behaves like Data.Map"
+    [ testPropertyN 500 "insert" (insertHMDATA @Int),
+      testPropertyN 500 "delete" (deleteHMDATA @Int),
+      testPropertyN 500 "union" (unionHMDATA @Int),
+      testPropertyN 500 "intersection" (intersectHMDATA @Int),
+      testPropertyN 500 "lookup" (lookupHMDATA @Int),
+      testPropertyN 500 "withoutKeys" (withoutHMDATA @Int),
+      testPropertyN 500 "restrictKeys" (restrictHMDATA @Int),
+      testPropertyN 500 "lookupMin" (minHMDATA @Int),
+      testPropertyN 500 "lookupMax" (maxHMDATA @Int),
+      testPropertyN 500 "splitLookup" (splitHMDATA @Int),
+      testPropertyN 500 "minViewWithKey" minViewHMDATA,
+      testPropertyN 500 "maxViewWithKey" maxViewHMDATA
+    ]
+
+-- ====================================================
+-- Generating random KeyMaps
+
+makeKeys :: Int -> Int -> [Key]
+makeKeys seed cnt = loop (mkStdGen seed) cnt []
+  where
+    loop g i !ans =
+      if i <= 0
+        then ans
+        else case genKey g of (!key, !g2) -> loop g2 (i - 1) (key : ans)
```

Comment:
With `Uniform` instance this becomes as easy as:
```
```suggestion
makeKeys :: Int -> Int -> [Key]
makeKeys seed cnt = runStateGen_ (mkStdGen seed) (uniformListM cnt)
```

Reply:
done